### PR TITLE
Elasticsearch improvements

### PR DESCRIPTION
**Pull request by dwradcliffe**
Still WIP

Review please @karmi :smile:

### CODE REVIEW

File: app/views/searches/show.html.erb [hunk 1]
```diff
@@ -7,6 +7,17 @@
   </div>
 <% end %>
 
+<% if @gems && @gems.size < 1 && @gems.respond_to?(:response) && (suggestions = @gems.response.response['suggest']) && suggestions.present? && suggestions['suggest_name'].present? && suggestions['suggest_name'][0]['options'].length > 0  %>
```
Comment (dwradcliffe) [inline app/views/searches/show.html.erb]: @karmi This seems really terrible. I can move this to a helper but is there a better way?

File: app/models/concerns/rubygem_searchable.rb [hunk 2]
```diff
@@ -37,9 +40,13 @@ def as_indexed_json(_options = {})
              }
 
     mapping do
-      indexes :name, analyzer: 'rubygem'
-      indexes :indexed, type: 'boolean'
-      indexes :info, analyzer: 'english'
+      indexes :name, type: 'multi_field' do
+        indexes :name, analyzer: 'rubygem'
+        indexes :suggest, type: 'completion'
```
Comment (dwradcliffe) [inline app/models/concerns/rubygem_searchable.rb]: @karmi Is this right? I tried a few different ways and this seemed to work. I couldn't find good documentation about the syntax here.

File: app/views/searches/show.html.erb (hunk 1 above)
Comment (nateberkopec) [inline app/views/searches/show.html.erb]: seems like a situation tailor-made for hash/array dig or the new safe navigation operator. Helper for now seems more readable at least.

File: test/functional/searches_controller_test.rb [hunk 3]
```diff
@@ -69,6 +107,38 @@ class SearchesControllerTest < ActionController::TestCase
...
+      @brando.index_document
+      Rubygem.__elasticsearch__.refresh_index!
+      get :show, query: "sinatre", es: 'true'
+    end
+
+    should respond_with :success
+    should render_template :show
+    should "see sinatra on the page in the suggestions" do
+      page.assert_text('Maybe you mean')
+      # assert page.find('.search-suggestions').has_content?(@sinatra.name)
```
Comment (dwradcliffe) [inline test/functional/searches_controller_test.rb]: Maybe someone with more capybara experience can show me the right way to do this?

File: app/views/searches/show.html.erb (hunk 1 above)
Comment (karmi) [inline app/views/searches/show.html.erb]: Yes, this is definitely too silly :( I guess you lifted it from the [example application](https://github.com/elastic/elasticsearch-rails/blob/master/elasticsearch-rails/lib/rails/templates/index.html.dsl.erb#L50), where it doesn't look very well as well.

I guess this should be better supported by the gem itself, at least allowing `@gems.suggestions` as a safe method, to get rid of most of the defensivity. What do you think?

Comment (dwradcliffe) [inline app/views/searches/show.html.erb]: Yeah, that's what I was hoping you would say :)

File: app/models/concerns/rubygem_searchable.rb (hunk 2 above)
Comment (karmi) [inline app/models/concerns/rubygem_searchable.rb]: The `completion` type is more suited to the auto-complete feature. If we want to first add the "Did you mean?" functionality, we should be able to just use a regular analyzer (eg. `simple`, as in the [example](https://github.com/elastic/elasticsearch-rails/blob/master/elasticsearch-rails/lib/rails/templates/searchable.dsl.rb#L17), and then use the [`suggest`](https://github.com/elastic/elasticsearch-rails/blob/master/elasticsearch-rails/lib/rails/templates/searchable.dsl.rb#L209) DSL method in the query. I have to play with it with some real data, to see how it behaves.

File: test/functional/searches_controller_test.rb (hunk 3 above)
Comment (karmi) [inline test/functional/searches_controller_test.rb]: I'll look into this, David. I guess the functional test creates the gems and indexes them, so it's an integration test?

File: app/models/concerns/rubygem_searchable.rb [hunk 4]
```diff
@@ -37,9 +40,13 @@ def as_indexed_json(_options = {})
              }
 
     mapping do
-      indexes :name, analyzer: 'rubygem'
-      indexes :indexed, type: 'boolean'
-      indexes :info, analyzer: 'english'
+      indexes :name, type: 'multi_field' do
```
Comment (karmi) [inline app/models/concerns/rubygem_searchable.rb]: The `multi_field` option is actually useless in current Elasticsearch — but I'll have to revisit the example application first, then we can carry it over.

Comment (dwradcliffe) [inline app/models/concerns/rubygem_searchable.rb]: For some reason it didn't seem to work without `multi_field` but it's possible I did something else wrong too.

File: app/views/searches/show.html.erb [hunk 5]
```diff
@@ -7,6 +7,17 @@
...
+  <div class='search-suggestions'>
+    <p>
+      Maybe you mean
+      <%= suggestions.map { |k,v| v.first['options'] }.flatten.map {|v| v['text']}.uniq.map do |term|
```
Comment (karmi) [inline app/views/searches/show.html.erb]: I think this should be also better supported by the gem itself — I just blindly return the ES JSON here, and it's not the ideal structure for an expressive environment like Rails...

Comment (dwradcliffe) [inline app/views/searches/show.html.erb]: Yeah, I'm sure there's a better way. This mapping is probably pretty standard.

**Pull request merged**